junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/JavaClassSelector.java

```diff
@@ -13,36 +13,60 @@
 import static org.junit.platform.commons.meta.API.Usage.Experimental;
 
 import org.junit.platform.commons.meta.API;
+import org.junit.platform.commons.util.PreconditionViolationException;
+import org.junit.platform.commons.util.ReflectionUtils;
 import org.junit.platform.commons.util.ToStringBuilder;
 import org.junit.platform.engine.DiscoverySelector;
 
 /**
  * A {@link DiscoverySelector} that selects a {@link Class} so that
  * {@link org.junit.platform.engine.TestEngine TestEngines} can discover
- * tests or containers based on Java classes.
+ * tests or containers based on classes.
+ *
+ * If a java {@link Class} is provided, the selector will return this
+ * {@link Class} and its class name accordingly. If the selector was
+ * created with a {@link Class} name, it will tries to lazy load the
+ * {@link Class} only on request.
  *
  * @since 1.0
  * @see org.junit.platform.engine.support.descriptor.JavaClassSource
  */
 @API(Experimental)
 public class JavaClassSelector implements DiscoverySelector {
 
-	private final Class<?> javaClass;
+	private final String className;
+	private Class<?> javaClass;
+
+	JavaClassSelector(String className) {
+		this.className = className;
+	}
 
 	JavaClassSelector(Class<?> javaClass) {
+		this.className = javaClass.getName();
 		this.javaClass = javaClass;
 	}
 
 	/**
+	 * Get the selected {@link Class} name.
+	 */
+	public String getClassName() {
+		return this.className;
+	}
+
+	/**
 	 * Get the selected Java {@link Class}.
 	 */
 	public Class<?> getJavaClass() {
+		if (this.javaClass == null) {
+			this.javaClass = ReflectionUtils.loadClass(this.className).orElseThrow(
+				() -> new PreconditionViolationException("Could not load class with name: " + this.className));
+		}
```

Comment:
Do you mean race conditions? It might but it's not expected that two threads will call it. And if so, it's lazy loading resulting in the same state. So no troubles if two threads calculate the result.